frontend/src/utils/generate-warnings.ts

```diff
@@ -75,6 +90,473 @@ export function produceWarnings(schedule: Schedule): IWarning[] {
   return warnings;
 }
 
+/**
+ * Identify a list of satisfied requirements given a major and a schedule.
+ * @param schedule the schedule to check requirements for.
+ * @param major the major to check requirements against.
+ */
+export function produceSatisfiedReqGroups(
+  schedule: Schedule,
+  major: Major
+): string[] {
+  let reqWarnings: IRequirementGroupWarning[] = produceRequirementGroupWarning(
+    schedule,
+    major
+  );
+  let reqGroups: Set<string> = new Set(major.requirementGroups);
+
+  for (const unsatisfiedWarning of reqWarnings) {
+    reqGroups.delete(unsatisfiedWarning.requirementGroup);
+  }
+
+  return Array.from(reqGroups);
+}
+
+/**
+ * Identify unsatisfied requirements given a major and a schedule.
+ * @param schedule the schedule to check requirements for.
+ * @param major the major to check requirements against.
+ */
+export function produceRequirementGroupWarning(
+  schedule: Schedule,
+  major: Major
+): IRequirementGroupWarning[] {
+  // holds courses that are currently on the schedule.
+  const taken: Map<string, number> = new Map<string, number>();
+  const coursesUsed: Set<string> = new Set<string>();
+
+  //add courses from the schedule to a Map: string => number (produceCourseCode => creditHours)
+  schedule.years.sort((n1, n2) => n1 - n2);
+  for (const yearNum of schedule.years) {
+    const year = schedule.yearMap[yearNum];
+
+    // add courses, term by term.
+    addCoursesToMap(year.fall, taken);
+
+    addCoursesToMap(year.spring, taken);
+
+    // if is summer full, only check summer I.
+    // shouldn't matter, summerII would just be empty.
+    if (year.isSummerFull) {
+      addCoursesToMap(year.summer1, taken);
+    } else {
+      addCoursesToMap(year.summer1, taken);
+      addCoursesToMap(year.summer2, taken);
+    }
+  }
+
+  let requirementGroups: string[] = major.requirementGroups;
+  let res: IRequirementGroupWarning[] = [];
+  for (const name of requirementGroups) {
+    let requirementGroup: IMajorRequirementGroup =
+      major.requirementGroupMap[name];
+    // todo: if req group is RangeSection or contains a ICourseRange process after all other requirement groups.
+    // this is because these reqs can be satisfied by a wide range of courses, and you don't want it using up
+    // a course that is need to satisfy a more constrained requirement.
```

Comment:
No I have not yet implemented this. But i think this okay because we don't really hit this edge case for any of the majors that we support currently. I will implement this in a future PR, hopefully soon.